solr/core/src/java/org/apache/solr/handler/component/ShardFieldSortedHitQueueWithSameShardCompareSkip.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.handler.component;
+
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.SortField;
+
+/**
+ * Used by distributed search to merge results.
+ * If two documents are on the same shard, their orderInShard is used for sorting instead of comparing their sort values.
+ * This shortcut does only work if we are not reRanking the results.
+ */
+public class ShardFieldSortedHitQueueWithSameShardCompareSkip extends ShardFieldSortedHitQueue {
```

Comment:
Have you considered not having a `ShardFieldSortedHitQueueWithSameShardCompareSkip` class but for the existing `ShardFieldSortedHitQueue` class to support variant behaviour e.g. something like this

```
public class ShardFieldSortedHitQueue ... {

+ private final boolean orderInShard;

public ShardFieldSortedHitQueue(SortField[] fields, int size, IndexSearcher searcher) {
  this(fields, size, searcher, true /* useOrderInShard */);
}

public ShardFieldSortedHitQueue(SortField[] fields, int size, IndexSearcher searcher, boolean useOrderInShard) {
  ...
  this.useOrderInShard = useOrderInShard;
}

...
-   if (docA.shard == docB.shard) {
+   if (docA.shard == docB.shard && this.useOrderInShard) {
...

}
```

noting that the 3-arg constructor signature maintains existing i.e. backwards compatible behaviour for the class even if perhaps `QueryComponent` with the changes always uses the new 4-arg constructor.

Reply:
We had the discussion yesterday. I am with you on this approach and will talk to the others again later.

Reply:
They agreed to changing it and I did so in 66e4f39.

Reply:
We've implemented this solution.